### PR TITLE
Tp/smartersegmentation

### PR DESCRIPTION
**Pull request by tom-parker**
cc @commuterjoy @wheresrhys

### CODE REVIEW

File: server/controllers/allocate.js [hunk 1]
```diff
@@ -11,18 +11,20 @@ module.exports = function(req, res, next) {
...
+				return allocateTests(res.locals.tests, res.locals.user, req);
+			}
+			return false;
+		})
+		.then((allocation) => {
+
+			if (allocation) {
+				res.set('ft-allocation-id', res.locals.user.uuid);
+			} else {
+				metrics.count('allocation.failed.tests', 1);
```
Comment (wheresrhys) [inline server/controllers/allocate.js]: this might end up duplicating metrics: for every failed.uuid there'll be a failed.tests

File: package.json [hunk 2]
```diff
@@ -9,7 +9,9 @@
     "ft-next-express": "^15.0.0",
     "n-health": "^1.2.0",
     "node-uuid": "^1.4.3",
-    "seedrandom": "^2.3.11"
+    "redis": "^0.12.1",
```
Comment (commuterjoy) [inline package.json]: you don't use this I think?

File: server/filters/referrer.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+const redis = require('then-redis');
+const url = require('url');
+const redisgreen = url.parse(process.env.REDIS_URL);
```
Comment (commuterjoy) [inline server/filters/referrer.js]: it's worth throwing an error if REDIS_URL isn't set as sending _undefined_ to `url.parse` with throw an error - i.e. would be easier to mandate REDIS_URL.

File: server/filters/referrer.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		metrics.count(`allocation.filter.${name}.redis.add.failed`, 1);
+	});
+}
+
+function getUserFromTest(uuid, name) {
+	return db.sismember(name, uuid).then((value) => {
+		return !!value;
+	})
+	.catch(() => {
+		metrics.count(`allocation.filter.${name}.redis.get.failed`, 1);
```
Comment (commuterjoy) [inline server/filters/referrer.js]: Perhaps name this after the redis instruction? Eg, `allocation.filter.${name}.redis.sismember.failed`

File: server/lib/allocate-tests.js [hunk 5]
```diff
@@ -18,24 +19,48 @@ module.exports = function(tests, user) {
...
+			return (filterReturnValues.every((value) => value)) ? test : false;
+		})
+		.catch(() => {
+			metrics.count(`allocation.filter.${test.name}.nofilterimplementation`, 1);
+		});
 	});
 
-	return allocation.join(',');
+	return Promise.all(filteredTests).then((vals) => {
+		const allocation = vals.filter((value) => value).map((test) => {
```
Comment (commuterjoy) [inline server/lib/allocate-tests.js]: Maybe some code comments here - it's quite hard to follow. Might just be late :)

File: test/allocate-tests.test.js [hunk 6]
```diff
@@ -15,87 +31,215 @@ const test_6 = { name: 'InRangeA', abTestSetup: {offset:90, range: 10} };
...
+		.then((val) => {
+			expect(val).to.equal(false);
+			done();
+		});
 	});
 
-	it('Should not allocate users if there are no active tests', function () {
-		expect(allocate({flagsWithABTests: []}, user)).to.equal(false);
+	it('Should not allocate users if there are no active tests', function (done) {
+		allocate({flagsWithABTests: []}, user)
```
Comment (commuterjoy) [inline test/allocate-tests.test.js]: yes, promise interface is nicer. thanks for changing it.

File: test/allocate-tests.test.js [hunk 7]
```diff
@@ -15,87 +31,215 @@ const test_6 = { name: 'InRangeA', abTestSetup: {offset:90, range: 10} };
...
 
-	it('Should allocate a user to a single test', function () {
-		expect(allocate({
+	it('Should allocate a user to a single test', function (done) {
+		allocate({
 			flagsWithABTests: [test_1],
 			anonymousTests: [test_1]
-		}, user)).to.deep.equal('foo:control');
+		}, user).then((val) => {
+			expect(val).to.deep.equal('foo:control');
```
Comment (commuterjoy) [inline test/allocate-tests.test.js]: We don't really need to `deep.equal` strings, just `equal` will do

**Comment by commuterjoy** [comment]
+1 looks good I think, aside from comments.

**Comment by commuterjoy** [comment]
and need to fix the CI build
